src/charm.py

```diff
@@ -53,21 +58,107 @@ def _on_peer_relation_changed(self, event):
             ] = event.relation.data[event.app]["MYSQL_ROOT_PASSWORD"]
             logger.info("Storing MYSQL_ROOT_PASSWORD in StoredState")
 
+    def _on_config_changed(self, _):
+        """This method handles the .on.config_changed() event"""
+        self._configure_pod()
+
+    # Handles start event
+    def _on_start(self, event):
+        """Initialize MySQL
+
+        This event handler is deferred if initialization of MySQL
+        fails. By doing so it is gauranteed that another
+        attempt at initialization will be made.
+        """
+
+        if not self.unit.is_leader():
+            return
+
+        if not self.mysql.is_ready():
+            msg = "Waiting for MySQL Service"
+            self.unit.status = WaitingStatus(msg)
+            logger.debug(msg)
+            event.defer()
+            return
+
+        self._on_update_status(event)
+        self._stored.mysql_initialized = True
+        self.unit.status = ActiveStatus()
+
+    # Handles update-status event
+    def _on_update_status(self, event):
+        """Set status for all units
+        Status may be
+        - MySQL is not ready,
+        - MySQL is not Initialized
+        - Unit is active
+        """
+        if not self.unit.is_leader():
+            self.unit.status = ActiveStatus()
+            return
+
+        if not self.mysql.is_ready():
+            status_message = "MySQL not ready yet"
+            self.unit.status = WaitingStatus(status_message)
+            return
+
+        if not self._stored.mysql_initialized:
+            status_message = "MySQL not initialized"
+            self.unit.status = WaitingStatus(status_message)
+            return
+
+        self.unit.status = ActiveStatus()
+
+    def _provide_mysql(self) -> None:
+        if self._stored.mysql_initialized:
+            self.mysql_provider = MySQLProvider(
+                self, "database", self.provides
+            )
+            self.mysql_provider.ready()
+            logger.info("MySQL Provider is available")
+
+    @property
+    def mysql(self) -> MySQL:
+        """Returns MySQL object"""
+        mysql_config = {
+            "app_name": self.model.app.name,
+            "host": self.hostname,
+            "port": self.model.config["port"],
+            "user_name": "root",
+            "mysql_root_password": self._stored.mysql_setup[
+                "MYSQL_ROOT_PASSWORD"
+            ],
+        }
+        return MySQL(mysql_config)
+
+    @property
+    def hostname(self) -> str:
+        """Unit hostname"""
+        unit_id = self.unit.name.split("/")[1]
+        return "{0}-{1}.{0}-endpoints".format(self.model.app.name, unit_id)
```

Comment:
Are you sure about this ? I may have forgotten but I think the third format string should be model name (kubernetes namespace) and not application name. Perhaps you set the two to be the same. Worth checking!.

Reply:
I'll check it!
I created an issue related to this method #38

Reply:
In mongodb charm we are using:

```python
        return "{}-{}.{}-endpoints".format(self.app_name,
                                           id,
                                           self.app_name)
```

Reply:
Eeew. I need to check if I made the same mistake. Will do. Thanks. But if we are migrating to pebble and have to use IPs then I may not bother.